lib/SetupChecks/MailConnectionPerformance.php

```diff
@@ -0,0 +1,111 @@
+	public function __construct(
+		private IL10N $l10n,
+		private LoggerInterface $logger,
+		private ProvisioningMapper $provisioningMapper,
+		private MailAccountMapper $accountMapper,
+		private IMAPClientFactory $clientFactory,
+		private FolderMapper $folderMapper,
+	) {
+	}
+
+	public function getName(): string {
+		return $this->l10n->t('Mail Connection Performance');
+	}
+
+	public function getCategory(): string {
+		return 'mail';
+	}
+
+	public function run(): SetupResult {
+		// retrieve unique imap hosts for provisionings and abort if none exists
+		$hosts = $this->provisioningMapper->uniqueImapHosts();
+		if (empty($hosts)) {
+			return SetupResult::success();
+		}
+		// retrieve random account ids for each host
+		$accounts = [];
+		foreach ($hosts as $host) {
+			$accounts[$host] = $this->accountMapper->getRandomAccountIdsByImapHost($host);
+		}
+		// test accounts
+		$tests = [];
+		foreach ($accounts as $host => $collection) {
+			foreach ($collection as $accountId) {
+				$account = new Account($this->accountMapper->findById((int)$accountId));
+				$client = $this->clientFactory->getClient($account);
+				try {
+					$tStart = $this->microtime(true);
+					// time login
+					$client->login();
+					$tLogin = $this->microtime(true);
+					// time operation
+					$list = $client->listMailboxes('*');
+					$status = $client->status(key($list));
+					$tOperation = $this->microtime(true);
+					
+					$tests[$host][$accountId] = ['start' => $tStart, 'login' => $tLogin, 'operation' => $tOperation];
+				} catch (Throwable $e) {
+					$this->logger->warning('Error occurred while performing system check on mail account: ' . $account->getId());
+				} finally {
+					$client->close();
+				}
+			}
+		}
+		// calculate performance
+		$performance = [];
+		foreach ($tests as $host => $test) {
+			$tLogin = 0;
+			$tOperation = 0;
+			foreach ($test as $entry) {
+				[$start, $login, $operation] = array_values($entry);
+				$tLogin += ($login - $start);
+				$tOperation += ($operation - $login);
+			}
+			$performance[$host]['login'] = $tLogin / count($tests[$host]);
+			$performance[$host]['operation'] = $tOperation / count($tests[$host]);
+		}
+		// display performance test outcome
+		foreach ($performance as $host => $entry) {
+			[$login, $operation] = array_values($entry);
+			if ($login > 1) {
+				return SetupResult::warning(
+					$this->l10n->t('Slow mail service detected (%1$s) an attempt to connect to several accounts took an average of %2$s seconds per account', [$host, round($login, 3)])
+				);
+			}
+			if ($operation > 1) {
+				return SetupResult::warning(
+					$this->l10n->t('Slow mail service detected (%1$s) an attempt to perform a mail box list operation on several accounts took an average of %2$s seconds per account', [$host, round($operation, 3)])
+				);
+			}
+		}
+		return SetupResult::success();
+	}
+
+	public function microtime(bool $asFloat = false): mixed {
+		return microtime($asFloat);
+	}
```

Comment:
I'm assuming this was added to mock time in tests. I'd advise to refactor to either of the two common options:
1. Use `\OCP\AppFramework\Utility\ITimeFactory::getTime`. You will get less precision but for a comparison and check if the connection takes more than a second, you're probably fine.
2. Move this method into a new class.

The reason I don't like the current approach is that you are mocking an internal of a class while testing. That's an anti pattern to avoid. Only external classes should be mocked. Additionally, a public method is exposed, which serves no purpose otherwise.